include/circt/Dialect/HW/HWTypes.h

```diff
@@ -14,12 +14,31 @@
 #ifndef CIRCT_DIALECT_HW_TYPES_H
 #define CIRCT_DIALECT_HW_TYPES_H
 
+#include "circt/Dialect/HW/HWDialect.h"
 #include "circt/Support/LLVM.h"
 #include "mlir/IR/Types.h"
 
 namespace circt {
 namespace hw {
+class ArrayType;
 class TypedeclOp;
+
+/// Base class for all concrete HW dialect types.
+class HWType
+    : public ::mlir::Type::TypeBase<HWType, mlir::Type, mlir::TypeStorage> {
+  using Base::Base;
+
+public:
+  /// Support method to enable LLVM-style type casting.
+  static bool classof(Type type);
+
+  /// Predicates to support querying the concrete type.
+  bool isArrayType();
```

Comment:
Why not provide a templatized version of this and the get function below in addition to or instead of this? `bool isActually<T>()`?

Reply:
I was just following the method naming as described in the Clang internals doc, but it should be possible to template this by the desired type...

Reply:
After moving away from the Clang-style, these helpers are indeed templated.